controllers/cruisecontroltask_controller.go

```diff
@@ -70,121 +81,218 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
-		logDirsByBroker, err := scaler.LogDirsByBroker()
+	case tasksAndStates.NumActiveTasksByOp(banzaiv1alpha1.OperationRebalance) > 0:
+		logDirsByBroker, err := r.Scaler.LogDirsByBroker()
 		if err != nil {
 			log.Error(err, "failed to get list of volumes per broker from Cruise Control")
 			return requeueAfter(DefaultRequeueAfterTimeInSec)
 		}
-
 		brokerIDs := make([]string, 0)
-		for _, task := range tasksAndStates.GetActiveTasksByOp(OperationRebalanceDisks) {
-			if task == nil || task.IsDone() {
-				continue
-			}
+		unavailableBrokerIDs := make([]string, 0)
+		for _, task := range tasksAndStates.GetActiveTasksByOp(banzaiv1alpha1.OperationRebalance) {
+			brokerIDs = append(brokerIDs, task.BrokerID)
+			found := false
 			if onlineDirs, ok := logDirsByBroker[task.BrokerID][scale.LogDirStateOnline]; ok {
-				found := true
 				for _, dir := range onlineDirs {
-					if !strings.HasPrefix(strings.TrimSpace(dir), strings.TrimSpace(task.Volume)) {
-						found = false
+					if strings.HasPrefix(strings.TrimSpace(dir), strings.TrimSpace(task.Volume)) {
+						found = true
 					}
 				}
-				if found {
-					brokerIDs = append(brokerIDs, task.BrokerID)
+				if !found {
+					unavailableBrokerIDs = append(unavailableBrokerIDs, task.BrokerID)
 				}
-			} else {
-				task.Err = "log dir is not available in Cruise Control"
 			}
 		}
+		if len(unavailableBrokerIDs) > 0 {
+			log.Info("requeue as there are offline broker log dirs for rebalance", "brokerIDs", unavailableBrokerIDs)
+			// This requeue is not necessary because the cruisecontrloperation controller retry the errored task
+			return requeueAfter(DefaultRequeueAfterTimeInSec)
+		}
 
 		details := []interface{}{"operation", "rebalance disks", "brokers", brokerIDs}
-
-		result, err := scaler.RebalanceDisks(brokerIDs...)
+		cruiseControlOpRef, err := r.rebalanceDisks(ctx, instance, nil, brokerIDs)
 		if err != nil {
-			log.Error(err, "re-balancing disk(s) in Kafka cluster via Cruise Control failed", details...)
+			log.Error(err, "creating CruiseControlOperation for re-balancing disks has failed", details...)
 		}
-
-		for _, task := range tasksAndStates.GetActiveTasksByOp(OperationRebalanceDisks) {
+		for _, task := range tasksAndStates.GetActiveTasksByOp(banzaiv1alpha1.OperationRebalance) {
 			if task == nil {
 				continue
 			}
-			task.FromResult(result)
+			task.SetCruiseControlOperationRef(cruiseControlOpRef)
+			task.SetStateScheduled()
 		}
 	}
 
 	if err = r.UpdateStatus(ctx, instance, tasksAndStates); err != nil {
 		log.Error(err, "failed to update Kafka Cluster status")
 	}
-	return requeueAfter(DefaultRequeueAfterTimeInSec)
+
+	return reconciled()
+}
+
+func checkBrokersAvailability(scaler scale.CruiseControlScaler, brokerIDs []string) ([]string, error) {
+	states := []scale.KafkaBrokerState{scale.KafkaBrokerAlive, scale.KafkaBrokerNew}
+	// This can result NullPointerException when the capacity calculation is missing for a broker in the cruisecontrol configmap
```

Comment:
Can we fix that somehow? I guess we should if we can just in case. :D

Reply:
I have done it more nicely